### PR TITLE
chore: bump sor to 4.7.0 - feat: SOR level implementation to support 1) expanding mixed quoter v1 to L2s 2) support migrate mixed quoter v1 to mixed quoter v2 across chains

### PR DESCRIPTION
**Pull request by jsy1218**
this PR compliments https://github.com/Uniswap/smart-order-router/pull/746, and will be able to achieve:
1. routing-api to shadow sample mixed routing against mixed quoter v1 across L2s. There's no accuracy to compare, just make sure routing-api shadow mixed routing success rate is high - existing metrics in routing-api ON_CHAIN_QUOTE_PROVIDER_${this.EXACT_IN_METRIC}_TRAFFIC_TARGET_FAILED_CHAIN_ID_${this.chainId}

On L2, only Base has the MixedRouteQuoterV1 deployed (see [context](https://uniswapteam.slack.com/archives/C021SU4PMR7/p1715195416711929?thread_ts=1712077488.419319&cid=C021SU4PMR7)). I thought its due to the lack of FOT and exact out support. But its actually due to SOR only enables mixed routing on L1 https://github.com/Uniswap/smart-order-router/blob/66927c4e4f7e75138df48afab1c33a6c6d8d8f15/src/routers/alpha-router/alpha-router.ts#L2081.

Therefore in this PR, I will only enable shadow sampling against Base mixed routing on L2, but keep 100% mixed routing traffic on L1.

I also enhanced existing unit test coverages.

tests all passed on my local https://app.warp.dev/block/afa5Lrn5JUeIIscuDGm14B

Tested on my personal AWS account:

- [x]  L1 mainnet - I can see only V3 and Mixed, all hitting target quoter address. Because V3 is migrated from revert quoter to view-only quoter on L1, and MixedRouteQuoterV1 is live on L1. V4 is not deployed to L1.
Screenshot 2024-10-14 at 3 25 42 PM

- [x] Optimism - I can see only V3 hitting targer quoter address. V4 and MixedRouteQuoterV1 is not deployed to Optimism.
Screenshot 2024-10-14 at 3 31 37 PM

- [x] Arb -   I can see only V3 hitting targer quoter address. V4 and MixedRouteQuoterV1 is not deployed to Arb.
Screenshot 2024-10-14 at 3 33 32 PM

- [x] Polygon - I can see only V3 hitting targer quoter address. V4 and MixedRouteQuoterV1 is not deployed to Polygon.
Screenshot 2024-10-14 at 3 34 31 PM

- [x] Celo - I can see only V3 hitting targer quoter address. V4 and MixedRouteQuoterV1 is not deployed to Celo.
Screenshot 2024-10-14 at 3 35 14 PM

- [x] BNB - I can see only V3 hitting targer quoter address. V4 and MixedRouteQuoterV1 is not deployed to BNB.
Screenshot 2024-10-14 at 3 36 11 PM

- [x] AVAX - I can see only V3 hitting targer quoter address. V4 and MixedRouteQuoterV1 is not deployed to AVAX.
Screenshot 2024-10-14 at 3 37 28 PM

- [x] BASE - I can see only V3 hitting targer quoter address. V4 not deployed to BASE. MixedRouteQuoterV1 deployed, but sampling rate too low.
Screenshot 2024-10-14 at 3 38 21 PM

### CODE REVIEW

**Comment by jsy1218** [comment]
* **#868** <a href="https://app.graphite.dev/github/pr/Uniswap/routing-api/868?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @jsy1218 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: package.json [hunk 1]
```diff
@@ -87,7 +87,7 @@
     "@uniswap/router-sdk": "^1.14.0",
     "@uniswap/sdk-core": "^5.8.2",
     "@types/semver": "^7.5.8",
-    "@uniswap/smart-order-router": "4.6.1",
+    "@uniswap/smart-order-router": "/tmp/uniswap-smart-order-router-4.7.0.tgz",
```
Comment (jsy1218) [inline package.json:90]: this is intentional, to remember i have to bump SOR before merging this PR.

File: test/mocha/integ/handlers/quote/provider-migration/traffic-switch-quote-provider.test.ts [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import sinon, { SinonSpy } from 'sinon'
 import { metric } from '@uniswap/smart-order-router/build/main/util/metric'
 import { MetricLoggerUnit, RouteWithQuotes, USDC_MAINNET, WRAPPED_NATIVE_CURRENCY } from '@uniswap/smart-order-router'
-import { TrafficSwitchOnChainQuoteProvider } from '../../../../../../lib/handlers/quote/provider-migration/v3/traffic-switch-on-chain-quote-provider'
+import { TrafficSwitchOnChainQuoteProvider } from '../../../../../../lib/handlers/quote/provider-migration/traffic-switch-on-chain-quote-provider'
```
Comment (jsy1218) [inline test/mocha/integ/handlers/quote/provider-migration/traffic-switch-quote-provider.test.ts:4]: i think i will add more tests here

File: lib/handlers/quote/util/quote-provider-traffic-switch-configuration.ts [hunk 3]
```diff
@@ -8,106 +9,244 @@ export type QuoteProviderTrafficSwitchConfiguration = {
...
-        samplingExactOutPercentage: 0,
-      } as QuoteProviderTrafficSwitchConfiguration
+      switch (protocol) {
+        // only base as L2 has mixed quoter v1 deployed
+        case Protocol.MIXED:
+          return {
+            switchExactInPercentage: 0,
+            samplingExactInPercentage: 1,
+            switchExactOutPercentage: 0,
+            samplingExactOutPercentage: 1,
```
Comment (xrsv) [inline lib/handlers/quote/util/quote-provider-traffic-switch-configuration.ts:62]: should this be 1 or 100?

File: lib/handlers/quote/util/quote-provider-traffic-switch-configuration.ts [hunk 4]
```diff
@@ -8,106 +9,244 @@ export type QuoteProviderTrafficSwitchConfiguration = {
...
+            samplingExactInPercentage: 0,
+            switchExactOutPercentage: 0,
+            samplingExactOutPercentage: 0,
+          } as QuoteProviderTrafficSwitchConfiguration
+        default:
+          return {
+            switchExactInPercentage: 0.0,
+            samplingExactInPercentage: 0.1,
+            switchExactOutPercentage: 0.0,
+            samplingExactOutPercentage: 0.1,
```
Comment (xrsv) [inline lib/handlers/quote/util/quote-provider-traffic-switch-configuration.ts:247]: should this be 0.1 or 1?

**Review by xrsv — APPROVED** [review]
looks good - left a couple of comments

File: test/mocha/integ/handlers/quote/provider-migration/traffic-switch-quote-provider.test.ts (hunk 2 above)
Comment (jsy1218) [reply test/mocha/integ/handlers/quote/provider-migration/traffic-switch-quote-provider.test.ts:4]: done. I assert that modified metrics can get invoked.

File: lib/handlers/quote/util/quote-provider-traffic-switch-configuration.ts (hunk 3 above)
Comment (jsy1218) [reply lib/handlers/quote/util/quote-provider-traffic-switch-configuration.ts:62]: 1 actually. 100% is 100 in this file.

File: lib/handlers/quote/util/quote-provider-traffic-switch-configuration.ts (hunk 4 above)
Comment (jsy1218) [reply lib/handlers/quote/util/quote-provider-traffic-switch-configuration.ts:247]: 0.1, meaning 0.1%. it's the same as it has been at 0.1% https://github.com/Uniswap/routing-api/pull/868/files/a86acba2a09920a8377cbef7e6de2a636002a62a#diff-a714608191ad685cf372e11933218d41d038498910b953547a18d4266666bee5R230

**Pull request merged**